pkg/apis/sealed-secrets/v1alpha1/sealedsecret_expansion.go

```diff
@@ -247,17 +247,37 @@ func NewSealedSecret(codecs runtimeserializer.CodecFactory, pubKey *rsa.PublicKe
 	return s, nil
 }
 
-// Unseal decrypts and returns the embedded v1.Secret.
-func (s *SealedSecret) Unseal(codecs runtimeserializer.CodecFactory, privKeys map[string]*rsa.PrivateKey) (*v1.Secret, error) {
-	boolTrue := true
-	smeta := s.GetObjectMeta()
-
+func decrypt(privKeys map[string]*rsa.PrivateKey, valueBytes []byte, smeta metav1.Object) ([]byte, error) {
 	// This will fail to decrypt unless the same label was used
 	// during encryption.  This check ensures that we can't be
 	// tricked into decrypting a sealed secret into an unexpected
 	// namespace/name.
 	label := labelFor(smeta)
 
+	plaintext, err := crypto.HybridDecrypt(rand.Reader, privKeys, valueBytes, label)
+	if err == nil {
+		return plaintext, nil
+	}
+
+	// Sometimes it happens that users encrypt the secret with a scope and then later change the scope.
+	// Let's not give them a helpless error message "cannot decrypt the key" and instead do some work
+	// on their behalf: try to see if there is any other scope that would have worked.
+	// Performing this work is safe since the user could (and would) do it anyway to troubleshoot why it doesn't decrypt.
+	scopes := []SealingScope{StrictScope, NamespaceWideScope, ClusterWideScope}
+	for _, s := range scopes {
+		label := EncryptionLabel(smeta.GetNamespace(), smeta.GetName(), s)
+		if _, err := crypto.HybridDecrypt(rand.Reader, privKeys, valueBytes, label); err == nil {
+			return nil, fmt.Errorf("secret item encrypted with scope %q, being decrypted in scope %q", s.String(), SecretScope(smeta).String())
```

Comment:
Wouldn't it be worth wrapping the original error here? (again, it depends on possible errors, in particular, whether reaching this line means it's definitely because of the scope, or whether there can be other intermittent errors or similar).

Reply:
But more importantly, I don't think you want to shadow `err` here, as if the error is unrelated to scope (and so all the checks for decrypting with other scopes here fail), you'll have overwritten the original error which will be returned below, I think?